job/deployer.go

```diff
@@ -120,14 +147,104 @@ func (d *deployer) cleanPerNamespace(ctx context.Context, namespaceSpec models.N
-	var enrichedJobSpecs []models.JobSpec
-	for _, jobSpec := range jobSpecs {
-		hooks := d.dependencyResolver.FetchHookWithDependencies(jobSpec)
-		if len(hooks) > 0 {
-			jobSpec.Hooks = hooks
+func (d *deployer) enrichJobSpecs(ctx context.Context, jobSpecs []models.JobSpec, deploymentProjectSpec models.ProjectSpec) error {
+	jobsByDestination := models.JobSpecs(jobSpecs).GroupJobsByDestination()
+
+	jobIDResourceDependenciesMap, err := d.getJobIDResourceDependenciesMap(ctx, deploymentProjectSpec, jobsByDestination)
+	if err != nil {
+		return err
+	}
+
+	for i, job := range jobSpecs {
+		targetJobSpec := job
+		if err := d.enrichWithStaticDependencies(ctx, &targetJobSpec, deploymentProjectSpec); err != nil {
+			return fmt.Errorf("error while enriching jobspec %d with static dependencies: %w", targetJobSpec.ID, err)
+		}
+
+		resourceDependencies := jobIDResourceDependenciesMap[job.ID]
+		d.enrichWithResourceDependencies(&targetJobSpec, deploymentProjectSpec, resourceDependencies)
+
+		d.enrichWithHookDependencies(&targetJobSpec)
+
+		jobSpecs[i] = targetJobSpec
+	}
+	return nil
+}
+
+func (d *deployer) enrichWithStaticDependencies(
+	ctx context.Context,
+	jobSpec *models.JobSpec,
+	deploymentProjectSpec models.ProjectSpec,
+) error {
+	projectJobSpecRepo := d.projectJobSpecRepoFactory.New(deploymentProjectSpec)
+	staticDependencies, err := d.dependencyResolver.ResolveStaticDependencies(ctx, *jobSpec, deploymentProjectSpec, projectJobSpecRepo)
+	jobSpec.Dependencies = staticDependencies
+	return err
+}
+
+func (*deployer) enrichWithResourceDependencies(
+	jobSpec *models.JobSpec,
+	deploymentProjectSpec models.ProjectSpec,
+	resourceDependencies []models.JobSpec,
+) {
+	if jobSpec.Dependencies == nil {
+		jobSpec.Dependencies = make(map[string]models.JobSpecDependency)
+	}
+	for _, dependency := range resourceDependencies {
+		dependencyType := models.JobSpecDependencyTypeIntra
+		if dependency.NamespaceSpec.ProjectSpec.ID.UUID() != deploymentProjectSpec.ID.UUID() {
+			dependencyType = models.JobSpecDependencyTypeInter
+		}
+		jobDependencySpec := dependency
+		jobSpec.Dependencies[dependency.Name] = models.JobSpecDependency{
+			Project: &jobDependencySpec.NamespaceSpec.ProjectSpec,
+			Job:     &jobDependencySpec,
+			Type:    dependencyType,
+		}
+	}
+}
+
+func (d *deployer) enrichWithHookDependencies(jobSpec *models.JobSpec) {
+	hooks := d.dependencyResolver.FetchHookWithDependencies(*jobSpec)
+	jobSpec.Hooks = hooks
+}
+
+func (d *deployer) getJobIDResourceDependenciesMap(ctx context.Context, deploymentProjectSpec models.ProjectSpec,
+	jobsByDestination map[string]models.JobSpec) (map[uuid.UUID][]models.JobSpec, error) {
+	jobSources, err := d.jobSourceRepository.GetAll(ctx, deploymentProjectSpec.ID)
+	if err != nil {
+		return nil, fmt.Errorf("error getting job sources for project id %s: %w", deploymentProjectSpec.ID.UUID(), err)
+	}
+
+	// populating the resource URN job map
+	resourceURNJobMap := make(map[string]*models.JobSpec)
```

Comment:
we can optimize populating the resourceURNJobMap for external projects by doing either of these:
- gather all of the external project sources that need to be fetched from table and make only single call (`GetByDestinations`), or
- fetch the extra job specs from the table, having the logic in the query itself.

Reply:
this is simplified by moving the complexity to the repository and only by selecting needed fields, not the whole job spec for each dependency.